pkg/gpu/context.go

```diff
@@ -46,6 +45,21 @@ type systemContext struct {
 
 	// procfsObj is the procfs filesystem object to retrieve process maps
 	procfsObj procfs.FS
+
+	// selectedDeviceByPIDAndTID maps each process ID to the map of thread IDs to selected device index.
+	// The reason to have a nested map is to allow easy cleanup of data when a process exits.
+	// The thread ID is important as the device selection in CUDA is per-thread.
+	// Note that this is the device index as seen by the process itself, which might
+	// be modified by the CUDA_VISIBLE_DEVICES environment variable later
+	selectedDeviceByPIDAndTID map[int]map[int]int32
+
+	// gpuDevices is the list of GPU devices on the system. Needs to be present to
+	// be able to compute the visible devices for a process
+	gpuDevices []nvml.Device
+
+	// visibleDevicesCache is a cache of visible devices for each process, to avoid
+	// looking into the environment variables every time
+	visibleDevicesCache map[int][]nvml.Device
```

Comment:
i guess this will also be refactored into a "deviceManager" struct with few other maps/arrays here. right?

Reply:
This one would likely go into the "process context" part as it's more linked to the process lifetime, but yes.